File: fast_factorization.rb
def factors(n)
  n_factors =[]
  (1..Math.sqrt(n)).select { |factor| n % factor == 0 }.each do |value|
    n_factors << n/value
    n_factors << value
  end
  n_factors.uniq.sort
end

p factors(20)
p factors(39)

File: fast_factorization.pl
#!/usr/bin/perl
use strict;
use warnings;
use POSIX;

sub factors {
  my $n = shift;
  my @list = ();
  for (my $i = 1; $i <= sqrt($n); $i++) {
    if ($n % $i == 0) {
      push(@list,($n/$i));
      push(@list,$i);
    }
  }
  return uniq(sort(@list));
}

sub uniq {
  my %seen;
  return grep { !$seen{$_}++ } @_;
}

my @test = factors(4);
print "@test\n";
